src/EditorFeatures/Core.Wpf/IWpfDifferenceViewerExtensions.cs

```diff
@@ -17,85 +19,71 @@ internal static class IWpfDifferenceViewerExtensions
-                // calculate required width and height) get a chance to run and we need to wait until
-                // this computation is complete. Once computation is complete, the width and height
-                // need to be set from the UI thread. We use ConfigureAwait(true) to stay on the UI thread.
-                await _taskCompletion.Task.ConfigureAwait(true);
-
-                // The following work must always happen on UI thread.
-                AssertIsForeground();
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+#pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task (containing method uses JTF)
+                await CalculateSizeAsync(cancellationToken);
+#pragma warning restore CA2007 // Consider calling ConfigureAwait on the awaited task
 
                 // We have the height and width required to display the inline diff snapshot now.
                 // Set the height and width of the IWpfDifferenceViewer accordingly.
                 _diffViewer.VisualElement.Width = _width;
                 _diffViewer.VisualElement.Height = _height;
             }
 
-            private void SnapshotDifferenceChanged(object sender, SnapshotDifferenceChangeEventArgs args)
+            private async Task<IProjectionSnapshot> GetInlineBufferSnapshotAsync(CancellationToken cancellationToken)
             {
-                // The following work must always happen on UI thread.
-                AssertIsForeground();
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (_diffViewer.DifferenceBuffer.CurrentInlineBufferSnapshot is { } snapshot)
+                {
+                    return snapshot;
+                }
 
-                // This event handler will only be called when the inline diff snapshot computation is complete.
-                Contract.ThrowIfNull(_diffViewer.DifferenceBuffer.CurrentInlineBufferSnapshot);
+                var completionSource = new TaskCompletionSource<IProjectionSnapshot>(TaskCreationOptions.RunContinuationsAsynchronously);
+                _diffViewer.DifferenceBuffer.SnapshotDifferenceChanged += HandleSnapshotDifferenceChanged;
 
-                // We can go ahead and calculate the required height and width now.
-                CalculateSize();
-            }
+                // Handle cases where the snapshot was set between the previous check and the event registration
+                if (_diffViewer.DifferenceBuffer.CurrentInlineBufferSnapshot is { } snapshot2)
+                    completionSource.SetResult(snapshot2);
 
-            private void CalculateSize()
-            {
-                // The following work must always happen on UI thread.
-                AssertIsForeground();
+                try
+                {
+                    return await completionSource.Task.WithCancellation(cancellationToken).ConfigureAwaitRunInline();
```

Comment:
can you doc the ConfigureAwaitRunInline bit?

Reply:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.awaitextensions.configureawaitruninline?view=visualstudiosdk-2019